### PR TITLE
[RSDK-4810] Integrate StopPlan and MoveOnGlobeNew with execution_id system

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-4810) [Other Ticket](https://viam.atlassian.net/browse/RSDK-5158)

Note: 
The internals of the state package could definitely be more intuitive / better layed out. That said, I feel pretty good about the API of the package & the tests so I figure we can improve the internals as needed in the future. This PR is currently blocking the expose paths to users project so I figured getting something up for review as soon as the API was in a good place was reasonable.

- create services/motion/builtin/state package to manage the state of the builtin motion service's executions, which supports starting executions, stopping executions by resource, stopping all executions from a motion/builtin/state struct, recording succeeded & failed replans, recording succeeded & failed ends to executions, recording stopped executions, and recording plan status updates, all such that data can be updated & retrieved efficiently.
- change motion/builtin.MoveOnGlobeNew to call state.StartExecution
- change motion/builtin.StopPlan to call state.StopExecutionByResource
- change motion/builtin.ListPlanStatuses to call state.ListPlanStatuses
- change motion/builtin.PlanHistory to call state.PlanHistory
- change services/motion/builtin to create a new motion/builtin/state on reconfigure & stop the previous state if one exists (to stop any executions from prior to reconfiguration)
- change the name of the motion/builtin lock to follow go & RDK conventions
- change all motion/builtin methods to take a read lock on the builtin struct. Otherwise, they are not threadsafe & bad things could happen if reconfigure gets called in the middle of one of the motion methods executing. After this change reconfigure & any other operation that takes a write lock on the builtin struct can't happen until all methods on the service have terminated & vice versa.
-  update tests to call `motion/builtin.Close()` so that they don't leak resources
- add motion.TerminalStateSet to describe the PlanStates which are terminal

Manual testing (LMK if you want the scripts I used):

https://github.com/viamrobotics/rdk/assets/5927876/c52f48f1-fb72-4618-8670-e9e2bd58e6cc

### CODE REVIEW

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:151]: check if state is nil

File: services/motion/motion.go
Comment (nicksanford) [inline services/motion/motion.go:37]: Type aliases

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:132]: If this is the first time viam-server is configuring the MotionService, it will create a new state & add it to the builtIn struct. 

Every subsequent reconfigure we will stop any executions from before the reconfigure & create a new state.

Comment (nicksanford) [reply services/motion/builtin/builtin.go:132]: https://viam.atlassian.net/browse/RSDK-5158

File: services/motion/builtin/builtin.go
Comment (nicksanford) [reply services/motion/builtin/builtin.go:148]: https://viam.atlassian.net/browse/RSDK-5158

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:167]: These `validateNotClosed` ensures that the motion service conforms to the expectations the RDK resource system expects, which is that all model methods should return errors if the model instance is closed.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:145]: closed tracks whether or not the motion service has been closed. Protected by `lock`

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:143]: made lock a RWMutex so that reads don't need to block each other.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:92]: Named return variables are a sharp edge in go which I don't think we should be using unless there is a good reason to: https://go.dev/tour/basics/7

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:141]: I moved the mutex above everything it protects, following the mutext convention.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:140]: closed tracks whether or not the model has been closed so that all methods can return errors if they have been closed, as expected by the RDK resource system.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [reply services/motion/builtin/builtin.go:141]: I changed the mutext to a RWMutex as that reads don't need to block each other & changed its name to `mu` (short for mutex) to conform to the convention elsewhere in the go ecosystem & RDK.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:166]: All these methods need to take the read lock on the mutex, otherwise, they are not thread-safe during reconfiguration.

File: services/motion/client_test.go
Comment (nicksanford) [inline services/motion/client_test.go:371]: I realized that it was unnecessary to use `uuid.NewUUID` which can return an error & that everywhere else in the RDK `uuid.New` is used. Swapping these function out removed a bunch of error cases from helper functions which didn't need to exist.

File: services/motion/builtin/builtin_test.go
Comment (nicksanford) [inline services/motion/builtin/builtin_test.go:434]: All these motionService.Close calls need to be added so that tests don't leak go routines.

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:169]: test that all methods obey closed

File: services/motion/motion.go
Comment (nicksanford) [inline services/motion/motion.go:114]: type alias

File: services/motion/builtin/builtin.go
Comment (nicksanford) [inline services/motion/builtin/builtin.go:37]: I'll remove this. Apparently We don't actually need to enforce returning errors when we are closed.

File: services/motion/builtin/state/state.go
Comment (tessavitabile) [inline services/motion/builtin/state/state.go:41]: [opt] As part of https://viam.atlassian.net/browse/RSDK-5146, we'll need to extend this to support MoveOnMap requests. Not sure if you want to think about how to do that now and/or name this so it's clear it's only for MoveOnGlobe.

File: services/motion/builtin/builtin.go
Comment (tessavitabile) [reply services/motion/builtin/builtin.go:37]: Oh, why did this change?

File: services/motion/builtin/builtin.go
Comment (tessavitabile) [reply services/motion/builtin/builtin.go:133]: If I understand correctly, reconfiguration will clear all plan history, which is not what we want.

File: services/motion/builtin/state/state.go
Comment (tessavitabile) [inline services/motion/builtin/state/state.go:124]: [opt] I think this will also need the destination, heading, etc. Not sure if you are planning to add that in this ticket, or a later ticket.

**Review by tessavitabile — COMMENTED** [review]
This is great!! Added a few comments. No need to wait for my sign-off.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:41]: I think this struct should be able to be used for MoveOnMap as well, we will just need to add some new fields to support it & have a constructor for move on map.

I think that work can be done when we start making move on map async.

Does that sound ok?

File: services/motion/builtin/builtin.go
Comment (nicksanford) [reply services/motion/builtin/builtin.go:133]: That is indeed what this does. I thought that Eliot wanted the motion service to behave in the same manner as `AlwaysRebuild` which would have blown away the plan history as well. We can't use `AlwaysRebuild` as the motion service has weak dependencies which results in it always being in a reconfigure loop. see: https://github.com/viamrobotics/rdk/pull/2946 where we tried to change the motion service to always rebuild & had to revert it as it broke the RDK.

This is the comment that indicated to me that deleting plan state history on reconfiguration is the desired behavior: https://viam.atlassian.net/browse/RSDK-5158?focusedCommentId=22394

@raybjork @tessavitabile can y'all please let me know what the behavior should be?

File: services/motion/builtin/builtin.go
Comment (nicksanford) [reply services/motion/builtin/builtin.go:37]: it changed b/c I misunderstood the contract between resources & the resource manager. I though (erroneously) that resources needed to gurantee that once their `Close()` method is called, all other resource interface methods (in this case the methods that implement the motion.proto) need to return errors.

After chatting with the netcode team, it turns out that this is actually not a requirement. As a result, I'm going to remove closed tracking, returning error on closed, & this error.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:124]: yep, I plan on adding that in this ticket when we implement the actual business logic of MoveOnGlobe: https://viam.atlassian.net/browse/RSDK-4814

File: services/motion/builtin/builtin.go
Comment (raybjork) [reply services/motion/builtin/builtin.go:133]: Yes Nick, I am also of the understanding that this is the behavior we agreed on

File: services/motion/builtin/builtin.go
Comment (raybjork) [inline services/motion/builtin/builtin.go:130]: When would this be nil?

Comment (nicksanford) [reply services/motion/builtin/builtin.go:130]: when it is called from the constructor: https://github.com/viamrobotics/rdk/blob/a4abb90482235b692a061d13660bc804a6f17a9d/services/motion/builtin/builtin.go#L86

Comment (raybjork) [reply services/motion/builtin/builtin.go:130]: Ah I see whats happening I think I read this PR in the wrong order.

File: services/motion/builtin/builtin.go
Comment (tessavitabile) [reply services/motion/builtin/builtin.go:127]: I think this is fine. It's a little at odds with what we scoped for ListPlanStatuses: https://github.com/viamrobotics/api/blob/8fc476f421f79d8ea1a8efdffdb670c3559e8b6b/proto/viam/service/motion/v1/motion.proto#L56-L58. But I think it makes sense, since either way navigation will need to retry a waypoint if the status for it is unavailable. I think we should just update that comment to say that statuses will be clear on reinitialization or reconfiguration.

File: services/motion/builtin/builtin.go
Comment (tessavitabile) [reply services/motion/builtin/builtin.go]: Eric had asked me to include that guidance in my talk on bugs and anti-patterns. It would be helpful if the netcode team would send out new guidance, since probably others are still working under that assumption.

File: services/motion/builtin/state/state.go
Comment (tessavitabile) [reply services/motion/builtin/state/state.go:41]: Yes

File: services/motion/builtin/state/state.go
Comment (tessavitabile) [reply services/motion/builtin/state/state.go:124]: sounds good

File: services/motion/builtin/state/state.go
Comment (biotinker) [inline services/motion/builtin/state/state.go:214]: To be sure I understand, limitations on this would be:
- if the resource is being moved by another motion service, this will return nil
- if the resource is being moved by calls directly to that resource, this will return nil
- if the resource's frame system parent is being moved by *this* motion service, and that plan is therefore taking into account where this resources' geometries are, this will still return nil.

Comment (nicksanford) [reply services/motion/builtin/state/state.go:214]: 1. True: A State is scoped to a given motion service so it has no knowledge of what another motoin service instance is doing.
2. True: If the move wasn't created via MoveOnGlobe then the State won't know about it & this function will return nil
3. True: I wasn't aware that was a requirement

File: services/motion/builtin/state/state.go
Comment (biotinker) [inline services/motion/builtin/state/state.go:550]: Why is this here?

File: services/motion/builtin/state/state.go
Comment (biotinker) [inline services/motion/builtin/state/state.go:592]: Temporary in the sense that it's deprecated and you're planning on getting rid of it soon?

Comment (biotinker) [reply services/motion/builtin/state/state.go:592]: Also, I don't think this needs to be public

**Review by biotinker — CHANGES_REQUESTED** [review]
Finished with first pass. I'm not 100% sure I've grokked all of the changes and will do another pass.

File: services/motion/builtin/state/state.go
Comment (biotinker) [reply services/motion/builtin/state/state.go:214]: I don't think (3) is a requirement, as rover-on-a-rover would be a pretty niche case, and arm-on-a-rover is something we've not gone out of our way to support yet.

**Review by raybjork — CHANGES_REQUESTED** [review]

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:54]: I don't think I can do that bc:
1. I need to be able to change the behavior of a specific execution, which means I need to add the test information to the execution struct itself
2. state_test.go is not part of the state package b/c it is testing the state api, not internals.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:198]: correct

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:180]: I don't think I understand. Could we chat about this tomorrow?

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:224]: Sorry, it should have been `cs` for componentState

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:362]: sorry, removed. Was from an earlier version.

File: services/motion/builtin/state/state.go
Comment (biotinker) [reply services/motion/builtin/state/state.go:390]: IMO `ManagedGo` should only be used where it's really necessary; it will, on a panic, re-call the `go()`'d to start again from scratch, which can have some really nasty side effects. It's not obvious from the name that that's what that does; executing half of a function, panicking, and then re-executing the first half of that same function.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:550]: b/c this is the string returned as the reason when the reason why the Plan failed is due to replanning

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:581]: so that I can trigger replan, plan success & plan failure from within the state_test.go

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:592]: yes I plan on getting rid of this code soon. If it is a blocker for merging I can try to come up with a different approach for being able to tests the behavior when replanning, plan success & plan failure occur.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:602]: if `newPlanFailReason` is nil it indicates that the plan succeeded.

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go]: done

File: services/motion/builtin/builtin.go
Comment (nicksanford) [reply services/motion/builtin/builtin.go:133]: yes, moved

File: services/motion/builtin/state/state.go
Comment (nicksanford) [reply services/motion/builtin/state/state.go:73]: removed

File: services/motion/motion.go
Comment (nicksanford) [reply services/motion/motion.go:107]: B/c the state package needs to understand which PlanStates are terminal. Given those types are defined by the motion package I figured the data structure that defines which instances of that type are terminal (i.e. mean that the Plan will never be updated) was better than a comment.

If you feel differently I can just make this a comment on the types & write out the if statements in the state package.

**Pull request closed without merging**
